metadata_service/proxy/atlas_proxy.py

```diff
@@ -124,6 +150,71 @@ def _get_table_entity(self, *, table_uri: str) -> Tuple[EntityUniqueAttribute, D
                                     .format(table_uri=table_uri))
 
+    def _get_user_entity(self, user_id: str) -> EntityUniqueAttribute:
+        """
+        Fetches an user entity from an id
+        :param user_id:
+        :return:
+        """
+        try:
+            return self._driver.entity_unique_attribute("User",
+                                                        qualifiedName=user_id)
+        except Exception as ex:
+            raise NotFoundException('(User {user_id}) does not exist'
+                                    .format(user_id=user_id))
+
+    def _create_reader(self, metadata_guid: str, user_guid: str, reader_qn: str) -> None:
+        """
+        Creates a reader entity for a specific user and table uri.
+        :param metadata_guid: Table's metadata guid
+        :param user_guid: User's guid
+        :param reader_qn: Reader qualifiedName
+        :return:
+        """
+        reader_entity = {
+            'typeName': 'Reader',
+            'attributes': {'qualifiedName': reader_qn,
+                           'isFollowing': True,
+                           'count': 0,
+                           'entityMetadata': {'guid': metadata_guid},
+                           'user': {'guid': user_guid}}
+        }
+        self._driver.entity_bulk.create(data={'entities': [reader_entity]})
+
+    def _get_reader_entity(self, table_uri: str, user_id: str) -> EntityUniqueAttribute:
+        """
+        Fetch a Reader entity from parsing table uri and user id.
+        If Reader is not present, create one for the user.
+        :param table_uri:
+        :param user_id: Qualified Name of a user
+        :return:
+        """
+        table_info = self._extract_info_from_uri(table_uri=table_uri)
+        reader_qn = '{}.{}.metadata.{}.reader@{}'.format(table_info.get('db'),
+                                                         table_info.get('name'),
+                                                         user_id,
+                                                         table_info.get('cluster'))
+
+        try:
+            reader_entity = self._driver.entity_unique_attribute(
+                "Reader", qualifiedName=reader_qn)
+            if not reader_entity.entity:
+                # Fetch the table entity from the uri for obtaining metadata guid.
+                table_entity, table_info = self._get_table_entity(table_uri=table_uri)
+                # Fetch user entity from user_id for relation
+                user_entity = self._get_user_entity(user_id)
+                # Create reader entity with the metadata and user relation.
+                self._create_reader(table_entity.entity[self.ATTRS_KEY][self.METADATA_KEY][self.GUID_KEY],
+                                    user_entity.entity[self.GUID_KEY], reader_qn)
+                # Fetch reader entity after creating it.
+                reader_entity = self._driver.entity_unique_attribute(self.READER_TYPE, qualifiedName=reader_qn)
```

Comment:
nit: I'd usually put that in the create reader function, to help me with the testing and to keep the logic for reader creation in one place.

Reply:
It is nice to keep create reader entity agnostic to the logic of this specific api. :)